source/core/memory.cpp

```diff
@@ -57,69 +57,99 @@ std::size_t Memory::getByteCount() const {
   assert::that(address >= 0);
   assert::that(value.getSize() % _byteSize == 0);
   const std::size_t amount{value.getSize() / _byteSize};
   assert::that(amount >= 0);
-  // I could possibly implement an optional assertion check
+  assert::that(address + amount <= _byteCount);
+  _put(address, value, amount, ignoreProtection);
+}
+
+void Memory::tryPut(std::size_t address,
+                    const MemoryValue& value,
+                    bool ignoreProtection) {
+  std::size_t amount{value.getSize() / _byteSize};
   if (!ignoreProtection && isProtected(address, amount)) {
     // do not write anything
     return;
   }
-  assert::that(address + amount <= _byteCount);
-  _data.write(value, address * _byteSize);
+  amount = std::min(amount, _byteCount - address);
+  MemoryValue valueInsert = value.subSet(0, amount * _byteSize);
+  _put(address, valueInsert, amount, ignoreProtection);
   _wasUpdated(address, amount);
 }
 
-MemoryValue Memory::set(const std::size_t address,
+MemoryValue Memory::_set(std::size_t address,
+                         const MemoryValue& value,
+                         bool ignoreProtection) {
+  std::size_t amount{value.getSize() / _byteSize};
+  MemoryValue prev{_get(address, amount)};
+  _put(address, value, amount, ignoreProtection);
+  return prev;
+}
+
+MemoryValue Memory::set(std::size_t address,
                         const MemoryValue& value,
                         bool ignoreProtection) {
-  const std::size_t amount{value.getSize() / _byteSize};
+  std::size_t amount{value.getSize() / _byteSize};
   MemoryValue prev{get(address, amount)};
   put(address, value, ignoreProtection);
   return prev;
 }
 
+MemoryValue Memory::trySet(std::size_t address,
+                           const MemoryValue& value,
+                           bool ignoreProtection) {
+  std::size_t amount{value.getSize() / _byteSize};
+  MemoryValue prev{tryGet(address, amount)};
+  tryPut(address, value, ignoreProtection);
+  return prev;
+}
+
 void Memory::_appendMemoryValue(std::stringstream& strm,
                                 const MemoryValue& value) {
-  static const char hex[] = "0123456789ABCDEF";
-  bool zero = true;
-  // reversely iterate over all the bytes in value (high->low)
-  for (std::size_t l = (value.getSize() + 7) / 8; l-- > 0;) {
-    std::uint8_t byte = value.getByteAt(l * 8);
-    std::uint8_t upperChar = byte / 16;
-    std::uint8_t lowerChar = byte % 16;
-    // do not print if this is the first character && it's 0
-    if (!zero || upperChar != 0) {
-      strm.put(hex[upperChar]);
-      zero = false;
-    }
-    // do not print if this is the first character && it's 0
-    if (!zero || lowerChar != 0) {
-      strm.put(hex[lowerChar]);
-      zero = false;
-    }
-  }
-  // if no characters have been printed print 0
-  if (zero) {
-    strm.put('0');
-  }
+  std::string s = value.toHexString(false, true);
+  strm << s;
 }
 
 std::pair<std::map<std::string, std::size_t>,
```

Comment:
I bet I can write `using MapPair = ...`  or whatever faster than you can write this type. This seems literally why `typedefs`/`using` exists 😄